ConsumerAffairs/django-experiments
# PUMA-102 Conditional experiments - groundwork

# PUMA-102
 
## Story

Part of PUMA-27, this PR sets some groundwork, to split code into a few smaller PRs.

## Acceptance Criteria
  
* add "auto-enroll" checkbox to Experiments admin page
* add "conditional" inlines to Experiments admin page
    * create "description" and "template" fields
    * dummy-evaluat "template" text - true if contains the word true
* remove duplicate way to change experiment state on the admin page
    * make "state" field read-only
* for staff users, populate `window.ca_experiments` with JSON that contains data about auto-enrolled experiments
    * to be used later with a JS indicator to assist in verifying conditional templates (i.e. whether a conditional works as expected)
* do not affect the existing "enrol" logic, so that we can perform arbitrary complex tests in the future unrelated to these new features in the admin (i.e. behaviour of non-auto-enrollable experiments should not change in any way)
 
## How to test
 
Using a local instance of MatchingTool site.

Checkout branch `frnhr/PUMA-103-update_django_experiments`
Run `pip install -r requirements-local.txt`

### Enable the experiments

Enable these experients:
 * toll-free
 * iconless

To enable an experiment:
 * Open any wizard on FE (e.g. `/eyeglasses-online/`) - this actually creates experiments 
 * Open `/admin/experiments/experiment`
 * Click on an experiment
 * In `Relevant Goals` table select `user_reached_results_step`
 * Click blue `Enabled` tool button in the top-right section of the page
 * Click "Save"
 * Reload a wizard on FE - this populates other variates besides `control`, but experiment needs to be enabled for this to happen

> This workflow is rather unintuitive, and will change with PUMA-83
To enrol in a specific variate of an experiment:
 * Open `/admin/experiments/experiment`
 * Click on the experiment
 * Make sure it is enabled
    * if not, see above
 * Make sure there is that `control` is not the only available variate (columns in the topmost table)
     * if only `control` is available, reload a wizard on FE
         * some experiments (only `intermediary` ATM, I think) are not triggered immediately but instead on a later step of the wizard, so you might need to complete a wizard to trigger it
 * click on `control (N) - Click to enrol` link in the leader of the topmost table.
 * or click on the variate to enrol there (usually there is only one variate, but there can be more)

### Tests

- [x] Test 1 - Verify that existing experiments are working as before
1. Enable `iconless` experiment and enrol into `new` variate
2. Open `/eyeglasses-online/`, and verify that:
    * all steps are "iconless" style (i.e. options are in stacked rows instead of square boxes)
3. Enrol into `control` variate in `iconless` experiment
4. Open `/eyeglasses-online/`, and verify that:
    * all steps are "old" style (i.e. options are show in square boxes aligned left-to-right)

- [ ] Test 2 - Test auto-enrol
1. Open `/admin/experiments/experiment/iconless/change/`
    * the experiment should be enabled at this point from steps in Test 1
2. Mark `Auto enrol` checkbox
3. Add a `Conditional`
    * put anything in the `description` field
    * put `true` in `template` field
4. Save
5. Open `/eyeglasses-online/` on FE
6. Open developer console, type `ca_experiments`, and verify that:
    * returned value is like this:
       `{'auto_enroll': {'iconless': {'auto-enrolling': True, 'enrolled_variate': 'control_OR_new'}}}`
7. Open / reload any page in the admin, type `ca_experiments` in the dev console, and verify that:
    * returned value is not set
8. On the admin page for the experiment, unmark `Auto enrol`
9. Save
10. Open developer console, type `ca_experiments`, and verify that:
    * returned value is like this:
       `{'auto_enroll': {}}`
11. On the admin page for the experiment, mark `Auto enrol`
12. Edit the `template` field so that it does not contain `true` (put anything else in)
13. Save
14. Open / reload `/eyeglasses-online/` on FE, type `ca_experiments` in the dev console, and verify that:
    * returned value is like this:
       `{'auto_enroll': {'iconless': {'auto-enrolling': False, 'enrolled_variate': 'SAME_AS_BEFORE'}}}`
15. Enrol in the OTHER variate (it's AJAX, no need to save)
16. Open / reload `/eyeglasses-online/` on FE, and verify that:
    * `ca_experiments` value is like this:
       `{'auto_enroll': {'iconless': {'auto-enrolling': False, 'enrolled_variate': 'THE_OTHER_VARIATE'}}}`
